Add a README for the project directory, remove unused apps (#683)
discord-python/site: pydis_site/README.md

# `pydis_site` project directory

This directory hosts the root of our **Django project**[^1], and is responsible
for all logic powering our website. Let's go over the directories in detail:

> - [`apps`](./apps) contains our **Django apps**. If you want to add your own API

Reviewer: Is this written assuming some prior knowledge of Django e.g. what an "app" means in the Django world? Otherwise a glossary footnote would be nicer, like you did for the project.

Author: Unfortunately the Django documentation doesn't have a glossary entry for applications, since it's mostly just a directory structure built on conventions (but created with `startapp`). I do think some prior Django knowledge is useful, but maybe we should expand on it regardless. How about something like this?

> `apps` contains our Django apps, which contain isolated functionality and are created using the `manage.py startapp` [linked to the django docs for that command] command.

Reviewer 2: > `apps` contain our Django apps, which are the building blocks that make up our Django project. A Django project must always consist of one or more apps, and these apps can be made completely modular and reusable across any Django project. In our project, each app controls a distinct part of our website, such as the API or our resources system. 
> 
> For more information on reusable apps, see the official Django tutorial, [which has a section on reusable apps](https://docs.djangoproject.com/en/4.0/intro/reusable-apps/).
> 
> To learn more about our specific apps, see the README inside the app folder itself.

Author: Thanks, that's a good one. See fee12689

  endpoint or new functionality to our homepage, that's the place to go.
  <!-- Each individual application also has its own README.md that you can click

>   through. -->

Reviewer: Why is this comment here if it's not visible when rendered?

Author: It's a placeholder for the documentation I plan to add later (#674), so it's commented out for when that has a PR. Or should we already add it now?

Reviewer 2: Just add it. We'll make it true soon.

Author: d5c3d0b3

- [`static`](./static) contains our **static files**, such as CSS, JavaScript,
  images, and anything else that isn't either content or Python code. Static
  files relevant for a specific application are put into subdirectories named
  after the application.

- [`templates`](./templates) contains our **Django templates**. Like with static
  files, templates specific to a single application are stored in a subdirectory
  named after that application. We also have two special templates here:

  - `404.html`, which is our error page shown when a site was not found.

>   - `500.html`, which is our error page shown in the astronomically rare case

Reviewer: Citation needed

Author: Mark, thank you for your concerns. Should we remove the "astronomically" or compare to other improbable events (e.g. winning the lottery, getting struck by lightning) instead?

Reviewer 2: hahaha. this is fine. I like the boldness.

the website:

- [`context_processors.py`](./context_processors.py), which contains custom
  *context processors* that add variables to the Django template context. To
  read more, see the [`RequestContext` documentation from
  Django](https://docs.djangoproject.com/en/dev/ref/templates/api/#django.template.RequestContext)

- [`settings.py`](./settings.py), our Django settings file. This mostly just
  parses configuration out of your environment variables, so you shouldn't need
  to edit it directly unless you want to add new settings.

- [`urls.py`](./urls.py), which configures our Django URL routing by installing
  our apps into the routing tree.

- [`wsgi.py`](./wsgi.py), which serves as an adapter for
  [`gunicorn`](https://github.com/benoitc/gunicorn),
  [`uwsgi`](https://github.com/unbit/uwsgi) or other application servers to run
  our application in production. Unless you want to test an interaction between
  our application and those servers, you probably won't need to touch this.

Note that for both `static` and `templates`, we are not using the default Django
directory structure which puts these directories in a directory per app (in our

> case, this would for example be ``pydis_site/apps/content/static/``).

Reviewer: This should probably go before line 26.

Author: Good point. See ecc0abb1

# `pydis_site` project directory

This directory hosts the root of our **Django project**[^1], and is responsible
for all logic powering our website. Let's go over the directories in detail:

- [`apps`](./apps) contains our **Django apps**. If you want to add your own API
  endpoint or new functionality to our homepage, that's the place to go.
  <!-- Each individual application also has its own README.md that you can click
  through. -->

- [`static`](./static) contains our **static files**, such as CSS, JavaScript,
  images, and anything else that isn't either content or Python code. Static
  files relevant for a specific application are put into subdirectories named

>   after the application.

Reviewer 2: ```suggestion
  after the application. For example, static files used by the Resources app go in `static/resources`.
```
Examples are nice, even when you've already explained something well.

Author: Would it be nice here to reference the app by its name in a codeblock such that the connotation between the static folder name and the application name is more clear?

Reviewer 2: I'm not sure what you're asking. show it?

Author: I don't know either. I think I meant something like

```suggestion
  after the application. For example, static files used by the `resources` app go in `static/resources`.
```

> - [`templates`](./templates) contains our **Django templates**. Like with static

Reviewer 2: ```suggestion
- [`templates`](./templates) contains our **[Django templates](https://docs.djangoproject.com/en/dev/topics/templates/)**. Like with static
```
Maybe we can link to the relevant part of the Django docs here - which would be https://docs.djangoproject.com/en/4.0/topics/templates/

Author: Good one. I've updated the link to ùse the dev version instead.

  files, templates specific to a single application are stored in a subdirectory
  named after that application. We also have two special templates here:

  - `404.html`, which is our error page shown when a site was not found.

  - `500.html`, which is our error page shown in the astronomically rare case
    that we encounter an internal server error.

> 

Reviewer 2: ```suggestion
```
Only an 🥥 INSANE PERSON 🥥 would have three newlines here.

Note that for both `static` and `templates`, we are not using the default Django
directory structure which puts these directories in a directory per app (in our
case, this would for example be ``pydis_site/apps/content/static/``).

We also have a few files in here that are relevant or useful in large parts of
the website:

- [`context_processors.py`](./context_processors.py), which contains custom
  *context processors* that add variables to the Django template context. To
  read more, see the [`RequestContext` documentation from
  Django](https://docs.djangoproject.com/en/dev/ref/templates/api/#django.template.RequestContext)

- [`settings.py`](./settings.py), our Django settings file. This mostly just
  parses configuration out of your environment variables, so you shouldn't need

>   to edit it directly unless you want to add new settings.

Reviewer 2: Not happy with this explanation. `settings.py` controls all manner of crucial things, and it would be useful to give some examples of what they are rather than just saying you probably don't need it.

For example, we use it to set up logging, our connection to the database, which apps are run by the project, middleware, and to configure variables for `django-simple-bulma`, which determine which colors our project and which bulma extensions are available on our pages.

Author: I have stolen all of this. 69c88813

> - [`urls.py`](./urls.py), which configures our Django URL routing by installing
>   our apps into the routing tree.

Reviewer 2: ```suggestion
- [`urls.py`](./urls.py), the URL configuration for the project itself. Here we can forward certain URL paths to our different apps, which have their own `urls.py` files to configure where their subpaths will lead. These files determine _which URLs will lead to which Django views_.
```

# `pydis_site` project directory

This directory hosts the root of our **Django project**[^1], and is responsible
for all logic powering our website. Let's go over the directories in detail:

- [`apps`](./apps) contains our **Django apps**, which are the building blocks
  that make up our Django project. A Django project must always consist of one
  or more apps, and these apps can be made completely modular and reusable
  across any Django project. In our project, each app controls a distinct part
  of our website, such as the API or our resources system.  For more
  information on reusable apps, see the official Django tutorial, [which has a
  section on reusable
  apps](https://docs.djangoproject.com/en/dev/intro/reusable-apps/). To learn

>   more about our specific apps, see the README inside the app folder itself.

Reviewer 2: This part could really use a newline or two. It's a bit dense.

Author: hmm, wasn't aware markdown supported that. Let's do it.

Author: [8eca5c8](https://github.com/python-discord/site/pull/683/commits/8eca5c8ddeed9ef7b827da9f6e87df034c073d6d)

We also have a few files in here that are relevant or useful in large parts of
the website:

- [`context_processors.py`](./context_processors.py), which contains custom
  *context processors* that add variables to the Django template context. To
  read more, see the [`RequestContext` documentation from
  Django](https://docs.djangoproject.com/en/dev/ref/templates/api/#django.template.RequestContext)

- [`settings.py`](./settings.py), our Django settings file. This controls all
  manner of crucial things, for instance, we use it to configure logging, our
  connection to the database, which applications are run by the project, which
  middleware we are using, and variables for `django-simple-bulma` (which
  determines frontend colours & extensions for our pages).

- [`urls.py`](./urls.py), the URL configuration for the project itself. Here we
  can forward certain URL paths to our different apps, which have their own
  `urls.py` files to configure where their subpaths will lead. These files
  determine _which URLs will lead to which Django views_.

- [`wsgi.py`](./wsgi.py), which serves as an adapter for
  [`gunicorn`](https://github.com/benoitc/gunicorn),

>   [`uwsgi`](https://github.com/unbit/uwsgi) or other application servers to run

Reviewer: Do not dare to omit the serial comma.
```suggestion
  [`uwsgi`](https://github.com/unbit/uwsgi), or other application servers to run
```

Author: I'm sorry Mark. From the deepest roots of my cold, cold, heart, I have realized my grave mistake when missing out this comma. As a native English speaker, I am aware of the importance of placing serial commas, and I apologize that I have let you down.

Sometimes shit happens. Sorry for that

  Johannes
